Bug fixes and improvements

N4gtan opened the pull request:
Updated Actions

Created a release and debug cmake profile

Updated readme compile instruccions

Update example.xml for the new format

Avoid read funcionts being call past EoF

Fixed some compiling warnings

Added ability to maintain EDC Form 2 integrity via xml, fixes this #46 

Fix posible crash if end map iterator cannot be dereferenced

Dummy sectors now saves the subheader value to the xml, so it can regenerate the correct values for non-standart (0x00/0x20) sectors

Added a HACK to check if there is a post gap dummy and some warnings if the gap is a non-standart 150 one, fixes this #44 

Now outputs by default an .xml file with the name of the input is no -s arg is given, this is a QOL for easy drag & drop

Fixed calculation length of dir/file entries, fixes this #52 and this #23 

Removed some duplicated lines

Dirs entries are now sort before the calculation of LBA tree

Now the program supports read/update timestamps prior to 1970 in windows

Now files are extracted/built with the original timestamps instead of being offsetted previously and then re calculated at build time. This is for better compatibily with other CD authoring tools

N4gtan commented on src/dumpsxiso/cdreader.cpp:
```diff
@@ -38,7 +38,10 @@ bool cd::IsoReader::Open(const fs::path& fileName)
 	fseek(filePtr, 0, SEEK_SET);
 
-    fread(sectorBuff, CD_SECTOR_SIZE, 1, filePtr);
+    if (fread(sectorBuff, CD_SECTOR_SIZE, 1, filePtr) != 1) {
```
Fixes some compiling warnings

CookiePLMonster commented on src/dumpsxiso/main.cpp:
```diff
@@ -314,18 +333,23 @@ static EntryAttributes EstablishXMLAttributeDefaults(tinyxml2::XMLElement* defau
 	auto findMaxElement = [](const auto& map)
 	{
-		return std::max_element(map.begin(), map.end(), [](const auto& left, const auto& right) { return left.second < right.second; })->first;
+		if (map.size() > 0) {
```
```suggestion
		if (!map.empty()) {
```

CookiePLMonster commented on src/dumpsxiso/main.cpp:
```diff
@@ -314,18 +333,23 @@ static EntryAttributes EstablishXMLAttributeDefaults(tinyxml2::XMLElement* defau
 
 	// Write them out to the XML
 	defaultAttributesElement->SetAttribute(xml::attrib::GMT_OFFSET, defaultAttributes.GMTOffs);
+	defaultAttributesElement->SetAttribute(xml::attrib::FILE_FLAGS, (defaultAttributes.FFLAGS < 2) ? defaultAttributes.FFLAGS : defaultAttributes.FFLAGS - 2);
```
This `- 2` looks magic, what does it signify?

N4gtan replied:
This is for the hidden attribute flag, but I currently changed it to
if (defaultAttributes.HFLAG & 0x01) {
	defaultAttributesElement->SetAttribute(xml::attrib::HIDDEN_FLAG, 1);
}

CookiePLMonster commented on src/dumpsxiso/main.cpp:
```diff
@@ -839,6 +870,7 @@ void ParseISO(cd::IsoReader& reader) {
 
     cd::ISO_DESCRIPTOR descriptor;
 	auto license = ReadLicense(reader);
+	const char* xa_edc = CheckEDCXA(reader);
```
`tinyxml2::XMLElement::SetAttribute` taking a boolean exists, therefore you can modify `CheckEDCXA` to return `bool` and not a `"true"`/`"false"` string.

CookiePLMonster reviewed (changes requested):
I'm not entirely sure what exactly is mismatching, but with my existing test case XMLs produced from dumpsxiso without your PR, built images aren't matching Redump anymore.

![image](https://github.com/user-attachments/assets/0d7f8f5d-06c3-49a8-973a-1df40d22c78b)

It's probably a good idea to ensure compatibility with older XMLs.

N4gtan commented:
> I'm not entirely sure what exactly is mismatching, but with my existing test case XMLs produced from dumpsxiso without your PR, built images aren't matching Redump anymore.
> 
> ![image](https://private-user-images.githubusercontent.com/7947461/356867782-0d7f8f5d-06c3-49a8-973a-1df40d22c78b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjMzNzM1MDUsIm5iZiI6MTcyMzM3MzIwNSwicGF0aCI6Ii83OTQ3NDYxLzM1Njg2Nzc4Mi0wZDdmOGY1ZC0wNmMzLTQ5YTgtOTczYS0xZGY0MGQyMmM3OGIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDgxMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA4MTFUMTA0NjQ1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MWVjYTc4MDhiZDIxZTVmZTRhNmVlNjI0YzhmNmViNDQ5YTEzMGRjNWUzMDhmZWU1ZDIyNmJjOThhNzU0NThmNyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.8jiQZZ0_vkJb0j-7n4c7A4nVWdDFlJN2h1GXTnJFsDw)
> 
> It's probably a good idea to ensure compatibility with older XMLs. For reference, here's one from Advan Racing:

Strange because I leaved default presets if the new xml strings are missing, but I will look at it.

Also I'm planing to push another fixes and changes I have made after checking the problem with your old xml.

CookiePLMonster commented:
I checked my Advan Racing checksum afterwards and it seems like, while matching my test case, it seems to be including the audio track in the .bin file and therefore it still doesn't match Redump. It should be easier for you to test against my GT1 XML instead, as that has no audio tracks and the rebuilt image [matches Redump](http://redump.org/disc/5513/):

<details>

```xml
<iso_project image_name="Gran Turismo (Europe) (En,Fr,De,Es,It) (EDC).bin" cue_sheet="Gran Turismo (Europe) (En,Fr,De,Es,It) (EDC).cue">
    <track type="data">
        <identifiers system="PLAYSTATION" application="PLAYSTATION" volume="GRAN_TURISMO" publisher="SONY COMPUTER ENTERTAINMENT INC." creation_date="1998041412583300+36"/>
        <license file="test_gt1/license_data.dat"/>
        <default_attributes gmt_offs="36" xa_attrib="56" xa_perm="17" xa_gid="4000" xa_uid="3025"/>
        <directory_tree xa_perm="1365" xa_gid="0" xa_uid="0">
            <file name="SYSTEM.CNF" source="test_gt1/SYSTEM.CNF" type="data"/>
            <file name="SCES_009.84" source="test_gt1/SCES_009.84" type="data" xa_perm="273"/>
            <file name="GTOS.EXE" source="test_gt1/GTOS.EXE" type="data" xa_perm="273"/>
            <file name="GTMAIN.EXE" source="test_gt1/GTMAIN.EXE" type="data" xa_perm="273"/>
            <file name="GTMENU.EXE" source="test_gt1/GTMENU.EXE" type="data" xa_perm="273"/>
            <file name="GTEND.EXE" source="test_gt1/GTEND.EXE" type="data" xa_perm="273"/>
            <file name="REPLAY.DAT" source="test_gt1/REPLAY.DAT" type="data"/>
            <file name="SYSTEM.DAT" source="test_gt1/SYSTEM.DAT" type="data" xa_perm="273"/>
            <file name="GAMEFONT.DAT" source="test_gt1/GAMEFONT.DAT" type="data"/>
            <file name="MESSAGES.DAT" source="test_gt1/MESSAGES.DAT" type="data" xa_uid="1000"/>
            <file name="TITLE.DAT" source="test_gt1/TITLE.DAT" type="data"/>
            <file name="CAR.DAT" source="test_gt1/CAR.DAT" type="data" xa_perm="273" xa_uid="1000"/>
            <file name="CARCADE.DAT" source="test_gt1/CARCADE.DAT" type="data" xa_perm="273" xa_uid="1000"/>
            <file name="CARINF.DAT" source="test_gt1/CARINF.DAT" type="data"/>
            <file name="COURSE.DAT" source="test_gt1/COURSE.DAT" type="data" xa_perm="273" xa_uid="1000"/>
            <file name="BG.DAT" source="test_gt1/BG.DAT" type="data" xa_perm="273" xa_uid="1000"/>
            <file name="GAMEMENU.DAT" source="test_gt1/GAMEMENU.DAT" type="data" xa_uid="1000"/>
            <file name="PITMENU.DAT" source="test_gt1/PITMENU.DAT" type="data" xa_uid="1000"/>
            <file name="ARCADE.DAT" source="test_gt1/ARCADE.DAT" type="data"/>
            <file name="ARCADE2.DAT" source="test_gt1/ARCADE2.DAT" type="data"/>
            <file name="MENU_IMG.ARC" source="test_gt1/MENU_IMG.ARC" type="data" xa_uid="1000"/>
            <file name="MENU_RAW.ARC" source="test_gt1/MENU_RAW.ARC" type="data" xa_uid="1000"/>
            <file name="SOUND.DAT" source="test_gt1/SOUND.DAT" type="data"/>
            <file name="MUSIC.DAT" source="test_gt1/MUSIC.DAT" type="mixed" xa_perm="273"/>
            <file name="STREAM.DAT" source="test_gt1/STREAM.DAT" type="mixed" xa_uid="1000"/>
            <file name="MEASURE.QA" source="test_gt1/MEASURE.QA" type="data" xa_perm="273"/>
			<dummy sectors="150" type="2336"/>
        </directory_tree>
    </track>
</iso_project>
```

</details>

N4gtan commented:
> I checked my Advan Racing checksum afterwards and it seems like, while matching my test case, it seems to be including the audio track in the .bin file and therefore it still doesn't match Redump. It should be easier for you to test against my GT1 XML instead, as that has no audio tracks and the rebuilt image [matches Redump](http://redump.org/disc/5513/):

I just checked Advan Racing and I got a 1:1 copy, what I think is currently happening is releated to the date of files nor a XML thing, because I changed the behaviour of how they are extracted/stored. I think you alredy has the files previously with a date offsted to your local time and the changes I made does not modify the date at build time, instead modify it at extract time so it can be used with tools like CDmage preserving the same date.

N4gtan commented:
Now correctly generates the descriptor's subheader, year and file order for games built with newer Sony's mastering tool.

Changed the way directory records are written to a linear write instead of a recursive write.

Changed the way descriptor directories are written from Depth-First Search to Breadth-First Search.

The -pt command now support dummies entries.

Some minor fixes.

N4gtan commented on src/mkpsxiso/iso.cpp:
```diff
@@ -416,7 +409,7 @@ int iso::DirTreeClass::CalculateDirEntryLen() const
 		int dataLen = sizeof(cd::ISO_DIR_ENTRY);
 
-		dataLen += GetIDLength(entry.id);
+		dataLen += entry.id.length();
```
Here I have big dude, I don't know if no-XA also needs to be rounded to even, for now only the XA isos are rounded.

N4gtan commented:
> I checked my Advan Racing checksum afterwards and it seems like, while matching my test case, it seems to be including the audio track in the .bin file and therefore it still doesn't match Redump. It should be easier for you to test against my GT1 XML instead, as that has no audio tracks and the rebuilt image [matches Redump](http://redump.org/disc/5513/):

Tested this and the difference is the dummy **form2**, I deprecated the `type="2336"` and because of this it is defaulting to **form1**`(0x00)`. Currently I'm using `xa_uid` with the raw subheader value because there are games that use non-standar dummies like `0x89`. I really can't decide which variable to use, I just reused xa_uid so as not to create another new entry, but if u like I can reuse the `type` one, tell me your thoughts.

N4gtan commented:
I reverted back to use `type` entry for dummies, so it now correctly generates form2 sectors if `type="2336"` is set.

I think this can now be safely merged.

Also added some QoL in these new commits
If no output dir is given, extracts the files into a folder with the same name as the input file instead of spreading out all the files in the same folder.
-lba and -lbahead args now outputs a default file if no path is given.

And some minor fixes
Correct timecodes with the -lba argument.
Prevents a path argument from parsing another argument instead of a path.

CookiePLMonster reviewed (commented):
Looks good after the new changes, thank you for putting so much work into this!

However, two remarks that made me not approve it yet:
1. I'm still a little concerned about the file timestamp changes. While your changes make perfect sense in terms of usability, I am worried that breaking "working copies" for the existing users may not be a good idea. Do you think this change can be somehow annotated in the XML file so only "new" unpacks with `dumpsxiso` get the time offset changes?
2. Do you think it makes sense to make CMake configs more consistent by putting Release build outputs in the Release subdirectory? Right now those land in `build/` while Debug lands in `build/Debug`.

@spicyjpeg Do you have any additional comments?

N4gtan commented:
As for the first point, yes I can incluide it in the xml, however it wouldn't be better to preserve the old timestamp format as an argument? Or maybe both (xml and argument)? Tell me what do you think.

As for the second point, yes u are right, I noticed that behaviour recently when I switched to linux. When compiling with MSVS the binaries were output in a "Release" folder but no with GCC.

CookiePLMonster commented:
> As for the first point, yes I can incluide it in the xml, however it wouldn't be better to preserve the old timestamp format as an argument? Or maybe both (xml and argument)? Tell me what do you think.

I partly agree, but my point is that care should be taken to avoid cases where people have their existing XML files and working directories, then update mkpsxiso, and their built ISOs "break". There are people who rely on mkpsxiso providing them accurate rebuilds to reduce the delta patch sizes, so with an update like this their xdeltas will inflate in size only because they have not re-unpacked the original ISO, which IMO they should not have to do.

N4gtan commented:
Ok now the mkpsxiso timestamp generation should be retro-compatible.
Added the Release binary output folder and reverted back to st_mtime because I didn't know that ctime can't be updated on linux.

It would be good to also close all the issues mentioned in this PR if it is merged.

CookiePLMonster reviewed (approved):
> Ok now the mkpsxiso timestamp generation should be retro-compatible. Instead of adding a new string to the xml, I just did a check on one of the new features, so since these string are new they shouldn't be present in older xml's.

Excellent idea!

I ran my tests again with your changes and now they work as expected - HS and CTR have some other differences I haven't yet investigated, but those are **not** related to your work and appear in the current version too:
![image](https://github.com/user-attachments/assets/04af6309-7e56-4071-9dc6-210db23acffa)

Thank you for all the work! @spicyjpeg do you have any additional comments? If not I think we can merge it soon and make a new release.

@N4gtan I have one more favor to ask please. With how many iterations your PR had it's not trivial to infer what exactly is changed by only looking at the commit history - could you prepare an user-friendly changelog out of your change so we could include it in the release?

N4gtan commented:
Something like the following or you mean a more simplified log?

# Full changelog:
- Added support to read/update timestamps prior to 1970 on windows
- Files are now extracted/built with the original timestamps instead of being offsetted previously and then re calculated at build time. This is for better compatibily with other CD authoring tools
- Added the ability to maintain EDC Form 2 integrity via the `xa_edc` xml string for games without it
- Dummy sectors now saves the subheader submode value in integer form via the `type` xml string, so it can regenerate the correct values for non-standart (0x00/0x20) sectors
- Added the ability to maintain the files and folders "hidden" via the `h_flag` xml string
- Now correctly generates the descriptor's submode, year and file order for games built with newer (>=2003) Sony's mastering tool via the `new_type` xml string
- Prevented path arguments from parsing another argument instead of a path
- Removed some duplicated and unnecessary lines
- Fixed some compiling warnings
- Created a release and debug cmake profile
- Updated GitHub Actions
- Updated compile instruccions
- Updated `example.xml` with the new strings
- And some minor changes/fixes

### Changes only to dumpsxiso
- Avoid `cdreader.cpp` read funcionts being call past EoF
- Fixed posible crash if end map iterator cannot be dereferenced at `main.cpp` line 348
- Added a HACK to check if there is a post gap pause and some warnings if the gap is not a 150 sectors standart one
- Now outputs by default an `.xml` file with the name of the input is no `-s` arg is given (this also allow an easy drag & drop on windows)
- The `-pt` command now support dummies entries.
- If no output dir is given with the `-x` arg, extracts the files into a folder with the same name as the input file instead of spreading out all the files in the working directory

### Changes only to mkpsxiso
- Fixed calculation length of dir/file entries
- Dirs entries are now sort before the calculation of LBA tree
- Fixed Root date year for games build with the old (<2003) sony's mastering tool
- Changed the way directory records are written to a linear write instead of a recursive write
- Changed the way descriptor directories are written from Depth-First Search to Breadth-First Search
- `-lba` and `-lbahead` args now outputs a default file if no path is given
- Corrected timecodes with the `-lba` argument
- Corrected the CD-DA sector calculation for -lba's arguments

CookiePLMonster commented:
Looks good, I can work with that.

CookiePLMonster commented:
v2.10 with your changes is now out: https://github.com/Lameguy64/mkpsxiso/releases/tag/v2.10

Thanks a lot once again! That was a lot of work.
